drivers/serial/uart_nrfx_uarte.c

```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+				data->async->rx_timeout_slab);
+	}
+}
+
+static void endrx_isr(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+
+	if (!data->async->rx_enabled) {
+		return;
+	}
+	if (data->async->next_rx) {
+		nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STARTRX);
+	}
+	k_delayed_work_cancel(&data->async->rx_timeout_work);
+	size_t rx_len = nrf_uarte_rx_amount_get(cfg->uarte_regs);
+	struct uart_event evt = {
+		.type = UART_RX_RDY,
+		.data.rx.buf = data->async->cur_rx,
+		.data.rx.len = rx_len - data->async->rx_offset,
+		.data.rx.offset = data->async->rx_offset,
+	};
+	user_callback(dev, &evt);
+	data->async->last_rx_byte_cnt += evt.data.rx.len;
+	evt.type = UART_RX_BUF_RELEASED;
+	evt.data.rx_buf.buf = data->async->cur_rx;
+	user_callback(dev, &evt);
+	if (data->async->next_rx) {
+		data->async->cur_rx = data->async->next_rx;
+		data->async->rx_size = data->async->rx_next_size;
+	} else {
+		data->async->cur_rx = 0;
+		data->async->rx_size = 0;
+		if (IS_ENABLED(CONFIG_UARTE_NRF_HW_ASYNC)
+		    && data->async->hw_async) {
+			nrfx_ppi_channel_disable(data->async->rx_cnt.ppi);
+			nrfx_timer_clear(&data->async->timer);
+		} else {
+			data->async->rx_cnt.cnt = 0;
+		}
+		data->async->last_rx_byte_cnt = 0;
+		evt.type = UART_RX_DISABLED;
+		user_callback(dev, &evt);
+	}
+	data->async->last_rx_evt = data->async->last_rx_byte_cnt;
+	data->async->rx_next_size = 0;
+	data->async->next_rx = NULL;
+	data->async->rx_offset = 0;
+}
+
+static void rxto_isr(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	struct uart_event evt = {
+		.type = UART_RX_BUF_RELEASED,
+		.data.rx_buf.buf = data->async->cur_rx,
+	};
+	user_callback(dev, &evt);
+	data->async->cur_rx = NULL;
+	if (data->async->next_rx) {
+		evt.data.rx_buf.buf = data->async->next_rx;
+		user_callback(dev, &evt);
+		data->async->next_rx = NULL;
+	}
+	evt.type = UART_RX_DISABLED;
+	user_callback(dev, &evt);
+}
+
+static void txstopped_isr(struct device *dev)
+{
+
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	struct uart_event evt = {
+		.type = UART_TX_ABORTED,
+		.data.tx.buf = data->async->tx,
+		.data.tx.len = nrf_uarte_tx_amount_get(
+					get_dev_config(dev)->uarte_regs),
+	};
+	data->async->tx = NULL;
+	user_callback(dev, &evt);
+}
+
+static void endtx_isr(struct device *dev)
+{
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	u16_t amount = nrf_uarte_tx_amount_get(cfg->uarte_regs);
+
+	if (cfg->uarte_regs->CONFIG & UARTE_CONFIG_HWFC_Msk) {
+		k_delayed_work_cancel(&data->async->tx_timeout_work);
+	}
+	struct uart_event evt = {
+		.type = UART_TX_DONE,
+		.data.tx.buf = data->async->tx,
+		.data.tx.len = amount,
+	};
+	user_callback(dev, &evt);
+	data->async->tx = NULL;
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_TXSTOPPED);
+	nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STOPTX);
```

Comment:
Won't this result in `UART_TX_ABORTED` event notification? I guess `data->async->tx` should be checked in `txstopped_isr` to avoid this.

But wouldn't it be better to trigger this task before calling `user_callback` (and to change the order of event handling in `uarte_nrfx_isr` so that the `TXSTOPPED` event could be handled right after the `ENDTX` one, in the same ISR run)?